Parker bids physio

dpark6060 opened the pull request:
This update is (should be) the final MAJOR functionality update.  New features include:

- BIDS info is now in metadata without overwriting anything or copying in parent metadata
- File classification now pulls in parent modality and adds custom classifications based on its own properties

This adds to the old functionality:

- Reads zipped DICOM archive and extracts raw physio data in ".log" text . format 
- Optionally generates BIDS compliant .tsv files from these log files and generates additional timeseries' for any external triggers that may be present in the .log file, as well as a timeseries for the scanner's new volume trigger.
- allows the user to specify how to handle missing (skipped) data in the BIDS file:
- provides options for how to generate a new timeseries (i.e., generate a completely new timeseries with uniform spacing that may or may not align with the true samples, or only fill in timepoints where there are gaps, which retains all original points, but may still have slight (1/2 sample length) offsets.
- allows the user to specify how to interpolate the missing data points (fill with constant, interpolate with linear, spline, etc).
- Optionally performs QA on resulting BIDS file and provides graphical output for the user to view
- Optionally deletes the .log files and saves only the BIDS .tsv files
- Optionally creates the BIDS .json files.



Josh and Andy, don't worry, you don't need to re-review everything.  I've made your suggested changes (Unless I missed something specific you noticed about the exceptions, Josh). 

The major changes are the new "data_classifier" function, which was present before but unused.  It's now called on lines 281, 301, and 336 in the main run script.

dpark6060 committed:
Merge pull request #8 from flywheel-apps/master

Master to BIDS

dpark6060 committed:
added to do in run.py

dpark6060 committed:
First Attempt at setting file properties

dpark6060 committed:
HUGE update for ECG stuff.  File Classification work present but commented (in dev)

dpark6060 committed:
added ECG QC (Untested)

dpark6060 committed:
Complete rework.  Physio dict is now in object that does all the
functions.  It's super cool.

dpark6060 committed:
Removed redundant functions

dpark6060 committed:
Debugging for Docker Run

dpark6060 committed:
Passed first gear local tests

dpark6060 committed:
Added Doc strings plus successful instance run

dpark6060 committed:
Updated Gitignore for local files

dpark6060 committed:
Removed Unused imports

dpark6060 committed:
Update metadata for BIDS

dpark6060 committed:
Update metadata for BIDS

dpark6060 committed:
Dev1 commit for Bridge Center

dpark6060 committed:
Dev2 commit for Bridge Center

dpark6060 committed:
Dev2 commit for Bridge Center, fixed NaN output issue.

dpark6060 committed:
Physio file metadata test

dpark6060 committed:
Physio file metadata test

dpark6060 committed:
passed physio metadata test for BIDS data and physio labels.  Added
more extensive docstrings

dpark6060 committed:
Final changes from last code review
- added versions in dockerfile (Andy)
- added types to docstrings
- attempted to improve exception calling?
- yeah...

Commit:
Added Physio QC examples

Commit:
Removed junk files

Commit:
Removed output and unzipped dicom folders

Commit:
Attempt3

joshicola reviewed (commented):
QA figure (physio_qa_ECG.qa.png) has legend (blue, orange, green, red) in opposite order of plots (red, green, orange, blue).  Also, the purple and brown are very hard to see on the legend.

joshicola commented on .gitignore:
```diff
@@ -2,3 +2,8 @@
 *.iml
 *.xml
 utils/__pycache__
+/utils/cfmm2tar/
```
Add the .DS_Store to the .gitignore

joshicola commented on Dockerfile:
```diff
@@ -7,7 +7,7 @@
 
 # First start with a python runtime
-FROM flywheel/python:dcmtk.latest
+FROM flywheel/python:3.7.4-dcmtk3.6.4
```
I didn't even know these images where here!!!
Nice!

joshicola commented on Dockerfile:
```diff
@@ -28,9 +28,9 @@ RUN apt-get -qq update && apt-get -qq install -y \
-RUN rm -rf /root/.cache/pip && \
+
+RUN pip3 install -r requirements.txt && rm -rf /root/.cache/pip && \
     rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
```
For this to affect the final docker image size, it must occur in the same layer that created these files.
RUN apt-get...

joshicola commented on requirements.txt:
```diff
@@ -1,4 +1,5 @@
-flywheel-sdk
+flywheel-sdk==10.0.1
```
I was just talking to @kalebdfischer about this being of utmost importance!!!
Specifying the "release version" of specific libraries.

joshicola commented on run.py:
```diff
@@ -8,536 +7,141 @@
+    :type context: class: `flywheel.gear_context.GearContext`
+    :param output_dir: the directory to save the metadata file in.
+    :type output_dir: str
     """
```
Sphinx docstrings need a space between the description and the first parameter...otherwise they render horribly...

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+            output_dir = gear_context.output_dir
+            fu.set_environment(environ_json, gear_context.log)
+
+            # Set up Custom Dicionary to host user variables
```
"Dicionary" -> "Dictionary"

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+
+            # Extract some basic values and set up gear environment
+            output_dir = gear_context.output_dir
+            fu.set_environment(environ_json, gear_context.log)
```
This function returns the json/dictionary of the environment....but is not captured on line 222...  rather the the path to the json file...?

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+                gear_context.log.info('Removing .log files')
+                cmd = ['/bin/rm', output_dir, '*.log']
+
+                exec_command(cmd)
```
You are soooo busted.... This needs a context object!!!!

joshicola commented on run.py:
```diff
@@ -8,536 +7,141 @@
+            'ImageType does not indicate physio, however by virtue of the gear running successfully, we will assume physio type')
+
+    # Attempt to recover classification info from the input file
+    (config, modality, classification) = ([], None, {})
```
'config' never used.

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+            gear_context.custom_dict['environ'] = environ_json
+
+            # Create a 'dry run' flag for debugging
+            gear_context.custom_dict['dry-run'] = gear_context.config['Dry-Run']
```
Where possible, my "preference" has become to have the manifest name match the name I use in code as much as possible. 
Implement as you like....

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+            gear_context.custom_dict['dry-run'] = gear_context.config['Dry-Run']
+
+            # Set up a field for physio dictionaries (used for bidsifying and qc)
+            gear_context.custom_dict['physio-dicts'] = {}
```
physio-dicts is defined but never used.

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+
+            # Extract matches keyword for BIDS data format
+            matches = gear_context.get_input('DICOM_ARCHIVE')['object']['info']
+            if 'SeriesDescription' in matches.keys() and matches['SeriesDescription']:
```
If 'SeriesDescription' is not in .keys() this will FAIL.

```
if 'SeriesDescription' in matches.keys():
  if matches['SeriesDescription']:
```
will ensure flow.

joshicola commented on run.py:
```diff
@@ -8,536 +7,141 @@
+    try:
+        supported_filetypes = [".dcm", ".IMA"]
 
         fu.exists(dicom, gear_context.log, '.zip')
```
If the fu module has its own logger then passing the context log is unnecessary,.

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+            # Try to run the extract physio command:
+            run_physio_command = ['/usr/local/bin/extractCMRRPhysio', raw_dicom[0], output_dir]
+            gear_context.custom_dict['raw_dicom'] = raw_dicom[0]
+            exec_command(gear_context, run_physio_command)
```
`exec_command` should be surrounded by try/except.  It will ungracefully crash this gear (without exit status) if it should fail.

dpark6060 replied:
the whole main script is encapsulated in a try

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+            #########################################
+            # Get the output.log files and make the physio objects
+            # Set the output dir
+            physio_output_dir = gear_context.output_dir
```
Wuh? Same output directory as above?  Why not reuse? Or just do gear_context.output_dir?

I am one for 'clarity over conciseness' whenever possible.

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+
+            if not info_file:
+                gear_context.log.warning('No Info file found.  Failed Extraction?')
+                return
```
`return` will kick you out of the `main()` function without a sys.exit() return status.  We need those.

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+        # Catch any exceptions
+        except Exception as e:
+            gear_context.log.exception(e)
+            sys.exit(1)
```
OK.  I get that you are nesting everything else in this top-level try/except block.  
I don't understand how it is returning an exit status '0' without an explicit sys.exit(0)?

joshicola commented on run.py:
```diff
@@ -553,67 +157,213 @@ def main():
+                    gear_context.log.info("error updating metadata in log")
+                    gear_context.log.exception(e)
+
+                # Set some values that will be set by the config file
+                physio.fill_val = gear_context.config['Fill_Value']
+                physio.interp_method = gear_context.config['Interpolation_Method']
+                physio.tic_fill_strategy = gear_context.config["Missing_Data"]
```
Somehow, I feel like these could end up as options in your constructor... Same for `physio.set_info(info)`.

I am a big fan of **Encapsulation**...  Hiding things from the user that they do not need to see... 

But who actually reads gear-scripts besides us???

joshicola replied:
you could pass the `gear_context` to you physio object.... but then that would make your physio object dependent on the gear context.

dpark6060 replied:
Nailed it, that's what I'm trying to avoid.

joshicola reviewed (changes requested):
Good iteration.  Unfortunately, it would crash and burn under particular circumstances.... and never return the proper exit status.

One thing to consider is putting an `__init__.py` file under all module directories (e.g. `utils/__init__.py`).  This enables my code editor (PyCharm or VSCode) to capture some errors and warnings.

Commit:
Final changes from last code review
- added versions in dockerfile (Andy)
- added types to docstrings
- attempted to improve exception calling?
- Added spaces in docstrings
- fixed Josh's concerns (All critical ones, plus the majority of the others)
